From 2cc0af6054fd53bbcadcedc596a7e6236b6bd8f5 Mon Sep 17 00:00:00 2001
From: Zhou Guanghui <zhouguanghui1@huawei.com>
Date: Thu, 27 Oct 2022 11:52:04 +0000
Subject: [PATCH 063/595] [Huawei] mm: fix alloc CDM node memory for MPOL_BIND

Offering: HULK
ascend inclusion
category: bugfix
bugzilla: 187895

------------------------------------------------------

Memory can be allocated from a specified CDM node only when it is
allowed to apply for memory from the CDM node. Otherwise, memory
will be allocated from other non-CDM nodes that are not allowed by
th cpuset.

Fixes: c1434668e6772 ("mm: Be allowed to alloc CDM node memory for MPOL_BIND")

Signed-off-by: Zhou Guanghui <zhouguanghui1@huawei.com>
---
 mm/page_alloc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index c63195b728a4..d58ddd6e7f73 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -3874,7 +3874,8 @@ get_page_from_freelist(gfp_t gfp_mask, unsigned int order, int alloc_flags,
 			(alloc_flags & ALLOC_CPUSET) &&
 			!__cpuset_zone_allowed(zone, gfp_mask)
 #ifdef CONFIG_COHERENT_DEVICE
-			&& !(alloc_flags & ALLOC_CDM)
+			&& (!is_cdm_node(zone->zone_pgdat->node_id) ||
+			    !(alloc_flags & ALLOC_CDM))
 #endif
 		)
 				continue;
-- 
2.17.1

